File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = { ")" : "(", "]": "[", "}": "{"}

        for ch in s:
            if ch in mapping:
                if len(stack) == 0 or mapping[ch] != stack[-1]:
                    return False
                stack.pop()
            else:
                stack.append(ch)
        
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.java

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char current : s.toCharArray()) {
            if (current == '(' || current == '{' || current == '[') {
                stack.push(current);
            } else {
                if (stack.isEmpty()) return false;
                char lastOpened = stack.pop();
                if ((current == ')' && lastOpened != '(') ||
                    (current == '}' && lastOpened != '{') ||
                    (current == ']' && lastOpened != '[')) {
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}